### PR TITLE
Implement str.replace

### PR DESCRIPTION
**Pull request by S-YOU**
Implementing str.replace.
This will have some updates from #167, so this is WIP for now.

### CODE REVIEW

File: runtime/str.go [hunk 1]
```diff
@@ -453,6 +453,37 @@ func strNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
...
 }
 
+// strReplace returns a copy of the string s with the first n non-overlapping
+// instances of old replaced by new. If old is empty, it matches at the
+// beginning of the string. If n < 0, there is no limit on the number of
+// replacements.
+func strReplace(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+	// TODO: Support unicode replace.
+	expectedTypes := []*Type{StrType, StrType, StrType, ObjectType}
+	argc, n := len(args), -1
```
Comment (trotterdylan) [inline runtime/str.go]: It's better not to group variables that are unrelated since it's harder to find their definition.

File: runtime/str.go [hunk 2]
```diff
@@ -453,6 +453,37 @@ func strNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
...
+	if argc == 4 {
+		o, raised := IntType.Call(f, Args{args[3]}, nil)
+		if raised != nil {
+			return nil, raised
+		}
+		n = toIntUnsafe(o).Value()
+	}
+	s := toStrUnsafe(args[0]).Value()
+	old := toStrUnsafe(args[1]).Value()
+	new := toStrUnsafe(args[2]).Value()
```
Comment (trotterdylan) [inline runtime/str.go]: Try to avoid using names that collide with builtin names.

File: runtime/str.go [hunk 3]
```diff
@@ -453,6 +453,37 @@ func strNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
...
+	// CPython only, pypy supposed to be same as Go
```
Comment (trotterdylan) [inline runtime/str.go]: Can you elaborate on this comment?

File: runtime/str.go [hunk 4]
```diff
@@ -453,6 +453,37 @@ func strNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
...
+	if len(s) == 0 && len(old) == 0 && n >= 0 {
+		return NewStr("").ToObject(), nil
+	}
+	return NewStr(strings.Replace(s, old, new, n)).ToObject(), nil
```
Comment (trotterdylan) [inline runtime/str.go]: I think strings.Replace() may not be compatible with the Python version because when `old == ""`, strings.Replace: `matches at the beginning of the string and after each UTF-8 sequence` whereas in Python it will match after each *byte*. So if the string contains UTF-8 sequences, the results will differ (this is a good test case to have, as well).

File: runtime/str.go (hunk 3 above)
Comment (S-YOU) [inline runtime/str.go]: I haven't really tested on pypy, but looks like it is different.
http://bugs.python.org/issue28029
assert ''.replace('', 'x') == 'x'
assert ''.replace('', 'x', 1000) == ''

File: runtime/str.go (hunk 1 above)
Comment (S-YOU) [inline runtime/str.go]: Thank you, make sense.

File: runtime/str.go (hunk 2 above)
Comment (S-YOU) [inline runtime/str.go]: Sure, too much copying from CPython.

File: runtime/str.go (hunk 4 above)
Comment (S-YOU) [inline runtime/str.go]: Oh, I haven't think about UTF-8 yet, also unicode.

**Comment by S-YOU** [comment]
- Rebased, fix reviewed issues
- Added utf8 replace test cases, but not sure I am testing real issues.

File: runtime/str.go (hunk 4 above)
Comment (S-YOU) [inline runtime/str.go]: I am not very sure which utf8 sequence is best to test, but added some.

**Comment by trotterdylan** [comment]
Thanks for the changes! The issue I'm pointing out is that in Go strings.Replace() treats the input as a UTF-8 encoded string and if the old string is `""` then it will add a replacement in between each UTF-8 code point. E.g.:

```
package main

import (
	"fmt"
	"strings"
)

func main() {
	fmt.Println([]byte(strings.Replace("\xd0\xb2\xd0\xbe\xd0\xbb", "", "\x00", -1)))
}
```

outputs: [0 208 178 0 208 190 0 208 187 0]

vs. CPython

```
>>> [ord(c) for c in"\xd0\xb2\xd0\xbe\xd0\xbb".replace('', '\0')]
[0, 208, 0, 178, 0, 208, 0, 190, 0, 208, 0, 187, 0]
```

I don't think it's a good idea to use strings.Replace() because of these subtle incompatibilities.

**Comment by S-YOU** [comment]
Hmm, I see. Is that mean, need to implement by looping through bytes?

**Comment by trotterdylan** [comment]
Looks like there's a bytes replace: https://golang.org/pkg/bytes/#Replace

**Comment by S-YOU** [comment]
Thank you but it mention same thing, 
`If old is empty, it matches at the beginning of the slice and after each UTF-8 sequence`

The issue is only on empty old string? then I can just write a loop for that.

**Comment by trotterdylan** [comment]
It's probably worth looking at the go source code to make sure that's the
only peculiarity. But yeah maybe that's the best approach.

**Comment by S-YOU** [comment]
Updated. I checked both strings.Replace and bytes.Replace source. It appears that they doing special checking on len(old) == 0, so I added a loop to copy manually for len(old) case. 
	            if len(old) == 0 {
			if i > 0 {
				_, wid := utf8.DecodeRuneInString(s[start:])
				j += wid
			}
I was thinking to copy codes from go source, but a lot of code to copy on.

File: runtime/str.go [hunk 5]
```diff
@@ -453,6 +453,65 @@ func strNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
...
+	if argc == 4 {
+		n, raised = ToIntValue(f, args[3])
+		if raised != nil {
+			return nil, raised
+		}
+	}
+	s := toStrUnsafe(args[0]).Value()
+	old := toStrUnsafe(args[1]).Value()
+	sub := toStrUnsafe(args[2]).Value()
+	// CPython only, pypy supposed to be same as Go.
```
Comment (trotterdylan) [inline runtime/str.go]: I'm a little confused about the meaning of this comment. Could you try to make the message a little clearer?

File: runtime/str.go [hunk 6]
```diff
@@ -453,6 +453,65 @@ func strNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
...
+	if len(s) == 0 && len(old) == 0 && n >= 0 {
```
Comment (trotterdylan) [inline runtime/str.go]: I'm probably overlooking something but don't we always want to return "" when len(s) is zero?

File: runtime/str.go [hunk 7]
```diff
@@ -453,6 +453,65 @@ func strNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
...
+	if len(old) == 0 {
+		if n < 0 {
+			n = len(s) + 1
+		}
+		buf := bytes.Buffer{}
+		if n > 0 {
+			buf.WriteString(sub)
+			n--
+		}
+		for i := 0; i < len(s); i++ {
```
Comment (trotterdylan) [inline runtime/str.go]: Store len(s) as a temporary numBytes or something so it's not recomputed every iteration.

File: runtime/str.go [hunk 8]
```diff
@@ -453,6 +453,65 @@ func strNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
...
+	old := toStrUnsafe(args[1]).Value()
+	sub := toStrUnsafe(args[2]).Value()
+	// CPython only, pypy supposed to be same as Go.
+	if len(s) == 0 && len(old) == 0 && n >= 0 {
+		return NewStr("").ToObject(), nil
+	}
+	if n == 0 {
+		return NewStr(s).ToObject(), nil
+	}
+	if len(old) == 0 {
```
Comment (trotterdylan) [inline runtime/str.go]: Maybe switch the conditional to `len(old) > 0` and do the strings.Replace() call to exit the function early. That would mean this big block of code doesn't have to be indented and it's probably easier to read.

Comment (trotterdylan) [inline runtime/str.go]: Worth adding a comment about why we do all this extra work instead of using strings.Replace()

Comment (trotterdylan) [inline runtime/str.go]: Prefer `old == ""` to `len(old) == 0`

File: runtime/str.go [hunk 9]
```diff
@@ -453,6 +453,65 @@ func strNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
...
+			n--
+		}
+		for i := 0; i < len(s); i++ {
+			buf.WriteByte(s[i])
+			if n > 0 {
+				buf.WriteString(sub)
+				n--
+			} else {
+				i++
+				if i < len(s) {
```
Comment (trotterdylan) [inline runtime/str.go]: Can we do this branch after the loop? I think the code would read more clearly: do the required replacements, then write the remainder of the string.

I think things would be tidier if your loop condition dictated when you exit the loop:

```
i := 0
for n > 0 && i < numBytes {
  // .. replacement logic
  i++
  n--
}
```

As written it's not immediately obvious what the termination condition is.

File: runtime/str.go (hunk 6 above)
Comment (S-YOU) [inline runtime/str.go]: `"".replace("", "x") returns "x", but `"".replace("", "x",1)` return "" in CPython, so it is need imo.

File: runtime/str.go (hunk 5 above)
Comment (S-YOU) [inline runtime/str.go]: I'll try.

File: runtime/str.go (hunk 9 above)
Comment (S-YOU) [inline runtime/str.go]: Great idea, thank you.

File: runtime/str.go (hunk 7 above)
Comment (S-YOU) [inline runtime/str.go]: Sure

**Comment by S-YOU** [comment]
Updated those, Please take a look again.
